[Cutlass] Integrate EVT into CUDACPPScheduling

mlazos opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #152815
* #150907
* #151406
* __->__ #150906
* #152733

Previously merged:
* #151713
* #151405
* #150905
* #152306
* #152305


Allow epilogue nodes in cuda combined scheduling

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

mlazos committed:
[Cutlass] Integrate EVT into CUDACPPScheduling

Allow epilogue nodes in cuda combined scheduling

[ghstack-poisoned]

mlazos committed:
Update on "[Cutlass] Integrate EVT into CUDACPPScheduling"


Allow epilogue nodes in cuda combined scheduling

[ghstack-poisoned]

mlazos committed:
Update on "[Cutlass] Integrate EVT into CUDACPPScheduling"


Allow epilogue nodes in cuda combined scheduling

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

mlazos committed:
Update on "[Cutlass] Integrate EVT into CUDACPPScheduling"


Allow epilogue nodes in cuda combined scheduling

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

mlazos committed:
Update on "[Cutlass] Integrate EVT into CUDACPPScheduling"


Allow epilogue nodes in cuda combined scheduling

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

mlazos committed:
Update on "[Cutlass] Integrate EVT into CUDACPPScheduling"


Allow epilogue nodes in cuda combined scheduling

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

mlazos committed:
Update on "[Cutlass] Integrate EVT into CUDACPPScheduling"


Allow epilogue nodes in cuda combined scheduling

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

mlazos committed:
Update on "[Cutlass] Integrate EVT into CUDACPPScheduling"


Allow epilogue nodes in cuda combined scheduling

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

mlazos committed:
Update on "[Cutlass] Integrate EVT into CUDACPPScheduling"


Allow epilogue nodes in cuda combined scheduling

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

mlazos committed:
Update on "[Cutlass] Integrate EVT into CUDACPPScheduling"


Allow epilogue nodes in cuda combined scheduling

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

mlazos committed:
Update on "[Cutlass] Integrate EVT into CUDACPPScheduling"


Allow epilogue nodes in cuda combined scheduling

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

mlazos committed:
Update on "[Cutlass] Integrate EVT into CUDACPPScheduling"


Allow epilogue nodes in cuda combined scheduling

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

mlazos committed:
Update on "[Cutlass] Integrate EVT into CUDACPPScheduling"


Allow epilogue nodes in cuda combined scheduling

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

mlazos committed:
Update on "[Cutlass] Integrate EVT into CUDACPPScheduling"


Allow epilogue nodes in cuda combined scheduling

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

mlazos committed:
Update on "[Cutlass] Integrate EVT into CUDACPPScheduling"


Allow epilogue nodes in cuda combined scheduling

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

mlazos committed:
Update on "[Cutlass] Integrate EVT into CUDACPPScheduling"


Allow epilogue nodes in cuda combined scheduling

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

mlazos committed:
Update on "[Cutlass] Integrate EVT into CUDACPPScheduling"


Allow epilogue nodes in cuda combined scheduling

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

mlazos committed:
Update on "[Cutlass] Integrate EVT into CUDACPPScheduling"


Allow epilogue nodes in cuda combined scheduling

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

mlazos committed:
Update on "[Cutlass] Integrate EVT into CUDACPPScheduling"


Allow epilogue nodes in cuda combined scheduling

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

mlazos committed:
Update on "[Cutlass] Integrate EVT into CUDACPPScheduling"


Allow epilogue nodes in cuda combined scheduling

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

mlazos committed:
Update on "[Cutlass] Integrate EVT into CUDACPPScheduling"


Allow epilogue nodes in cuda combined scheduling

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

mlazos committed:
Update on "[Cutlass] Integrate EVT into CUDACPPScheduling"


Allow epilogue nodes in cuda combined scheduling

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

mlazos committed:
Update on "[Cutlass] Integrate EVT into CUDACPPScheduling"


Allow epilogue nodes in cuda combined scheduling

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

mlazos committed:
Update on "[Cutlass] Integrate EVT into CUDACPPScheduling"


Allow epilogue nodes in cuda combined scheduling

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

mlazos committed:
Update on "[Cutlass] Integrate EVT into CUDACPPScheduling"


Allow epilogue nodes in cuda combined scheduling

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

mlazos committed:
Update on "[Cutlass] Integrate EVT into CUDACPPScheduling"


Allow epilogue nodes in cuda combined scheduling

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

mlazos committed:
Update on "[Cutlass] Integrate EVT into CUDACPPScheduling"


Allow epilogue nodes in cuda combined scheduling

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

mlazos committed:
Update on "[Cutlass] Integrate EVT into CUDACPPScheduling"


Allow epilogue nodes in cuda combined scheduling

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

mlazos committed:
Update on "[Cutlass] Integrate EVT into CUDACPPScheduling"


Allow epilogue nodes in cuda combined scheduling

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

mlazos committed:
Update on "[Cutlass] Integrate EVT into CUDACPPScheduling"


Allow epilogue nodes in cuda combined scheduling

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

mlazos committed:
Update on "[Cutlass] Integrate EVT into CUDACPPScheduling"


Allow epilogue nodes in cuda combined scheduling

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

mlazos committed:
Update on "[Cutlass] Integrate EVT into CUDACPPScheduling"


Allow epilogue nodes in cuda combined scheduling

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

eellison commented on torch/_inductor/codegen/cuda/cuda_cpp_scheduling.py:
```diff
@@ -40,9 +45,32 @@ def is_cuda_cpp_template(node: BaseSchedulerNode) -> bool:
+                node2,  # type: ignore[arg-type]
+            )
+        elif self.is_cuda_cpp_fused_template(node1) and isinstance(
+            node2, SchedulerNode
```
Follow up: should handle FusedSchedulerNode

eellison commented on torch/_inductor/codegen/cuda/cuda_cpp_scheduling.py:
```diff
@@ -114,3 +148,93 @@ def codegen_template(
+    def _unwrap_epilogue_nodes(
+        fused_node: FusedSchedulerNode,
+    ) -> list[BaseSchedulerNode]:
+        nodes = list(fused_node.get_nodes())
```
nit: no need to create additional `list(...)`

eellison commented on torch/_inductor/codegen/cuda/cuda_cpp_scheduling.py:
```diff
@@ -114,3 +148,93 @@ def codegen_template(
+
+        if not isinstance(cuda_template_buffer, CUDATemplateBuffer):
+            return False
+        # if not cuda_template_buffer.template.can_fuse_epilogue:
+        #    # The used GEMM op does not support fusing epilogues
+        #    return False
```
?

eellison commented on torch/_inductor/codegen/cuda/cuda_cpp_scheduling.py:
```diff
@@ -114,3 +148,93 @@ def codegen_template(
+        # if not cuda_template_buffer.template.can_fuse_epilogue:
+        #    # The used GEMM op does not support fusing epilogues
+        #    return False
+        if not isinstance(additional_ir_node, ComputedBuffer):
+            return False
+        if not isinstance(additional_ir_node.data, Pointwise):
+            return False
```
https://github.com/pytorch/pytorch/blob/49cbe0ffe9925e8664d52628d55688fb4491e460/torch/_inductor/scheduler.py#L3645-L3651

Do you want these conditions ? Also the `why` for no fusion can be useful.

eellison commented on torch/_inductor/codegen/cuda/cuda_cpp_scheduling.py:
```diff
@@ -114,3 +148,93 @@ def codegen_template(
+        Args:
+            cuda_template_buffer : A CUDATemplateBuffer object representing the CUDA template and it's result buffer
+            epilogue_nodes : List[ir.Buffer]: The list of already fused epilogue nodes.
+            additional_node: The ir.Buffer node to be checked if it can be fused with the epilogue.
```
Nit: it's not an ir.Buffer..

eellison commented on torch/_inductor/codegen/cuda/cuda_cpp_scheduling.py:
```diff
@@ -40,9 +45,32 @@ def is_cuda_cpp_template(node: BaseSchedulerNode) -> bool:
+                [],
+                node2,  # type: ignore[arg-type]
+            )
+        elif self.is_cuda_cpp_fused_template(node1) and isinstance(
```
Typically we fuse all of the pointwise nodes first. Then fuse into templates. So its more likely that the second node is a FusedSchedulerNode and the cpp template is still just `cuda_cpp_template`..

eellison commented on torch/_inductor/codegen/cuda/cuda_cpp_scheduling.py:
```diff
@@ -114,3 +148,93 @@ def codegen_template(
+        self,
+        cuda_template_buffer: CUDATemplateBuffer,
+        epilogue_nodes: list[BaseSchedulerNode],
+        additional_node: BaseSchedulerNode,
```
nit: maybe - existing_epilogue_nodes / node_to_fuse ?

eellison commented on torch/_inductor/codegen/cuda/cuda_cpp_scheduling.py:
```diff
@@ -114,3 +148,93 @@ def codegen_template(
+        """
+        additional_ir_node = additional_node.node
+
+        if not isinstance(cuda_template_buffer, CUDATemplateBuffer):
```
feel like this should be an assertion since we already checked it..

eellison commented on torch/_inductor/codegen/cuda/cuda_cpp_scheduling.py:
```diff
@@ -114,3 +148,93 @@ def codegen_template(
+        # if not cuda_template_buffer.template.can_fuse_epilogue:
+        #    # The used GEMM op does not support fusing epilogues
+        #    return False
+        if not isinstance(additional_ir_node, ComputedBuffer):
+            return False
```
we just pass in the scheduler node above... is this ever true ?

mlazos replied:
Yes, we unpack the fused nodes, so we're guaranteed it's not a FusedSchedulerNode. Any individual SchedulerNodes should have internal nodes which are ComputedBuffers. (I convert to ir node at the top of the function)

mlazos replied:
I'll remove

eellison commented on torch/_inductor/codegen/cuda/cuda_cpp_scheduling.py:
```diff
@@ -114,3 +148,93 @@ def codegen_template(
+        if not isinstance(additional_ir_node.data, Pointwise):
+            return False
+        # We can fuse a Pointwise op that depends on the last fused epilogue node
+        # if any. If there is no epilogue node yet, it needs to depend on the template
```
not sure I follow `if any` ..

mlazos replied:
If there aren't any, it should depend on the mm.

eellison commented on torch/_inductor/codegen/cuda/cuda_cpp_scheduling.py:
```diff
@@ -114,3 +148,93 @@ def codegen_template(
+            return False
+
+        if len(epilogue_nodes) == 0:
+            if cuda_template_buffer.name not in additional_ir_node.get_read_names():
```
Feels like some of this should have already been handled in the simd scheduling.. Should this be an assertion ?

mlazos replied:
Yeah I can convert this to assertion

eellison commented on torch/_inductor/codegen/cuda/cuda_cpp_scheduling.py:
```diff
@@ -114,3 +148,93 @@ def codegen_template(
+                return False
+        else:
+            last_epilogue_node = epilogue_nodes[-1].node
+            assert isinstance(last_epilogue_node, ComputedBuffer)  # for mypy
```
Why do we have `epilogue_nodes: list[BaseSchedulerNode]` above then Comptued Buffer here ?

mlazos replied:
Yeah I think this was previously ready for review, but I've butchered it since, sorry! Thanks for taking a look though. The comments are helpful regardless. Will fix.

mlazos replied:
I removed this check, because it's possible the new node might depend on a node other than the last epilogue node.

eellison reviewed (commented):
Some of the conditions here should have already been checked in the non backend specific schedulilng..

add test ?

mlazos committed:
Update on "[Cutlass] Integrate EVT into CUDACPPScheduling"


Allow epilogue nodes in cuda combined scheduling

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

mlazos committed:
Update on "[Cutlass] Integrate EVT into CUDACPPScheduling"


Previously merged:
* #151713
* #151405
* #150905
* #152306
* #152305


Allow epilogue nodes in cuda combined scheduling

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

mlazos commented:
> Some of the conditions here should have already been checked in the non backend specific schedulilng..
> 
> add test ?

I'm going to add tests in the next PR, it needs the next change to work e2e

mlazos committed:
Update on "[Cutlass] Integrate EVT into CUDACPPScheduling"


Previously merged:
* #151713
* #151405
* #150905
* #152306
* #152305


Allow epilogue nodes in cuda combined scheduling

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

mlazos committed:
Update on "[Cutlass] Integrate EVT into CUDACPPScheduling"


Previously merged:
* #151713
* #151405
* #150905
* #152306
* #152305


Allow epilogue nodes in cuda combined scheduling

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

mlazos committed:
Update on "[Cutlass] Integrate EVT into CUDACPPScheduling"


Previously merged:
* #151713
* #151405
* #150905
* #152306
* #152305


Allow epilogue nodes in cuda combined scheduling

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]
